src/meltano/cli/utils.py

```diff
@@ -514,3 +516,37 @@ def check_dependencies_met(
     else:
         message = f"Dependencies not met: {'; '.join(messages)}"
     return passed, message
+
+
+class InstrumentedDefaultGroup(DefaultGroup):
+    """A variation of a DefaultGroup that instruments its invocation by updating the telemetry context."""
+
+    def invoke(self, ctx):
+        """Update the telemetry context and invoke the group."""
+        ctx.ensure_object(dict)
+        if ctx.obj.get("tracker"):
+            ctx.obj["tracker"].add_contexts(CliContext.from_click_context(ctx))
+        super().invoke(ctx)  # noqa: WPS608
+
+
+class InstrumentedGroup(click.Group):
+    """A click.Group that instruments its invocation by updating the telemetry context."""
+
+    def invoke(self, ctx):
+        """Update the telemetry context and invoke the group."""
+        ctx.ensure_object(dict)
+        if ctx.obj.get("tracker"):
+            ctx.obj["tracker"].add_contexts(CliContext.from_click_context(ctx))
+        super().invoke(ctx)  # noqa: WPS608
```

Comment:
@WillDaSilva so rather than what you had mentioned:

> Instead of building a CliContext from a Click context, we would have it built inside of InstrumentedGroup/InstrumentedCommand, and then stored in the Click context as a list under the key snowplow_contexts or something like that.

> And then with those contexts in place (by tracker.with_contexts(...)) the CLI events would fire, and would only contain the event (e.g. started, failed, inflight, etc.). All the other data would be in the contexts.

I kept the add_contexts pattern rather than adding to the click context, and then having `with_contexts` since we create the tracker early on in `cli.py` anyway. 

As long as we have a tracker obj - any instrumented Group or Command will end up managing/adding its own CliContext, and things firing CliEvents don't have to think about contexts at all. 

If we kaboom halfway through the cli resolution chain, we might at least still have *some* cli context's this way as well.